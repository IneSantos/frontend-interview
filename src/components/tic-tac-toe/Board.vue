<template>
  <div class="tictactoe-board">
    <div v-if="finished5rounds" class="status">
      <div v-if="stalemate" class="result stale">It's a draw!<br>Stalemate!</div>
      <div v-else class="result">Finished!<br><span class="winner">{{winner}}</span> wins!</div>
    </div>

    <div class="grid">
      <div v-for="(row, row_index) in 3" :key="row">
        <div v-for="(column, column_index) in 3" :key="column">
          <Cell :value="board[row_index][column_index]"  @click="playMove(row_index,column_index)"/>
        </div>
      </div>
    </div>

    <div class="score">
      <Player name="Player 1" :score="player1" class="score-info"/>
      <StopWatch ref="stopwatch" class="score-info"/>
      <Player name="Player 2" :score="player2" class="score-info"/>
    </div>
  </div>
</template>

<script>
import '@/styles/tic-tac-toe/board.scss';
import Player from './Player';
import Cell from './Cell';
import StopWatch from './StopWatch';

export default {
  name: 'TicTacToe',
  components: {
    Player,
    Cell,
    StopWatch
  },
  data() { 
    return {
      board:  [
        ['', '', ''],
        ['', '', ''],
        ['', '', '']
      ],
      player: "x",
      player1: 0,
      player2: 0,
      finished: false,
      stalemate: false,
      watchStarted: false,
      finished5rounds: false,
      winner: ''
    }
  },
  computed: {
    numOfgames () {
      return this.$store.getters.getNumOfGames;
    }
  },
  watch: {
    numOfgames: function () {
      this.finished5rounds =  this.numOfgames === 4;
      this.winner = this.player1 >  this.player2 ? "P1" : "P2";
      this.$store.dispatch('setGameHistory', this.winner);
    }
  },
  methods: {
      playMove(x, y) {
        if (this.finished || this.board[x][y] !== '') {
          return;
        }

        this.board[x][y] = this.player;
        this.$forceUpdate();

        if(!this.watchStarted){
          this.$refs.stopwatch.start();
          this.watchStarted = true;
          this.$store.dispatch('setGameStart', true);
        }

        this.finished = this.checkWinner() || this.checkStalemate();
        this.stalemate = this.checkStalemate();
        console.log("finished", this.finished);
        this.updateScore();
        this.nextPlayer();
        
      },
      nextPlayer: function() {
        this.player = this.player === "x" ? "o" : "x";
      },  
      checkStalemate: function() {
        return (
          !this.finished &&
          this.checkValuesPresent(this.board[0]) &&
          this.checkValuesPresent(this.board[1]) &&
          this.checkValuesPresent(this.board[2])
        );
      },
      checkWinner: function() {
        return (
          this.checkValues(this.board[0]) ||
          this.checkValues(this.board[1]) ||
          this.checkValues(this.board[2]) ||
          this.checkValues([this.board[0][0], this.board[1][0], this.board[2][0]]) ||
          this.checkValues([this.board[0][1], this.board[1][1], this.board[2][1]]) ||
          this.checkValues([this.board[0][2], this.board[1][2], this.board[2][2]]) ||
          this.checkValues([this.board[0][0], this.board[1][1], this.board[2][2]]) ||
          this.checkValues([this.board[0][2], this.board[1][1], this.board[2][0]])
        );
      },
      checkValues: function(values) {
        return this.checkValuesPresent(values) && this.checkValuesMatch(values);
      },
      checkValuesPresent: function(values) {
        return values[0] !== "" && values[1] !== "" && values[2] !== "";
      },
      checkValuesMatch: function(values) {
        return values[0] === values[1] && values[1] === values[2];
      },
      resetBoard: function() {
        this.board = [
            ['', '', ''],
            ['', '', ''],
            ['', '', '']
          ];
        this.finished = false;
        this.stalemate = false;  
        this.watchStarted = false;
      },
      updateScore: function() {
        if(!this.finished) return;

        this.player === "x" ? ++this.player1 :  ++this.player2;
        this.$store.dispatch('setWinsPlayer1', this.player1);
        this.$store.dispatch('setWinsPlayer2', this.player2);
        this.$store.dispatch('incNumOfGames');

        this.$refs.stopwatch.stop();
        this.$refs.stopwatch.reset();
        this.resetBoard();
      }

  }

}
</script>