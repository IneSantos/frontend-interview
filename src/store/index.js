export const store = {
    state: {
      hasgameStarted: false,
      numOfgames: 0,
      winsPlayer1: 0,
      winsPlayer2: 0,
      gameHistory: ['', '', '' , '', '']
    },
    mutations: {
      SET_GAME_STARTED(state, hasgameStarted) {
        state.hasgameStarted = hasgameStarted
      },
      INC_NUM_GAMES(state) {
        state.numOfgames++
      },
      SET_WINS_PLAYER_1(state, winsPlayer1) {
        state.winsPlayer1 = winsPlayer1
      },
      SET_WINS_PLAYER_2(state, winsPlayer2) {
        state.winsPlayer2 = winsPlayer2
      },
      SET_GAME_HISTORY(state, player) {
        state.gameHistory[state.numOfgames-1] = player;
      },
      RESET_GAME(state) {
        state.hasgameStarted = false;
      }
    },
    actions: {
      setGameStart(context, hasgameStarted){
        context.commit('SET_GAME_STARTED', hasgameStarted);
      },
      resetBoard(context){
        context.commit('RESET_BOARD')
      },
      incNumOfGames(context){
        context.commit('INC_NUM_GAMES');
      },
      setWinsPlayer1(context, winsPlayer1){
        context.commit('SET_WINS_PLAYER_1', winsPlayer1);
      },
      setWinsPlayer2(context, winsPlayer2){
        context.commit('SET_WINS_PLAYER_2', winsPlayer2);
      },
      setGameHistory(context, player){
        context.commit('SET_GAME_HISTORY', player);
      }
    },
    getters: {
      getHasGameStarted(state) {
        return state.hasgameStarted;
      },
      getNumOfGames(state) {
        return state.numOfgames;
      },
      getWinsPlayer1(state) {
        return state.winsPlayer1;
      },
      getWinsPlayer2(state) {
        return state.winsPlayer2;
      },
      getGameHistory(state) {
        return state.gameHistory;
      },
    }

}

