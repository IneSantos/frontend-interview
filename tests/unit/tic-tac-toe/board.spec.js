import { mount, shallowMount, createLocalVue } from '@vue/test-utils'
import Vuex from 'vuex'
import Board from '@/components/tic-tac-toe/Board.vue'
import { storeMocked } from '../../__mocks__/store.js'

const localVue = createLocalVue();
localVue.use(Vuex);

const store = new Vuex.Store(storeMocked);

describe('Board.vue', () => {
  it('renders default', () => {
    const wrapper = shallowMount( Board, { store, localVue });
    expect(wrapper).toMatchSnapshot();
  })

  it('renders finished', async () => {
    const wrapper = mount( Board, { store, localVue });
    await wrapper.setData({ finished5rounds: true});

    expect(wrapper).toMatchSnapshot();
  })

  it('Should call playMove when cell is clicked', async () => {
    const mockMethod = jest.spyOn(Board.methods, 'playMove');
    const wrapper = mount( Board, { store, localVue });

    await wrapper.findComponent({ ref: "cell-0-0" }).trigger('click');
    expect(mockMethod).toHaveBeenCalled();
  })

  it('Should call checkWinner when cell is clicked', async () => {
    const mockMethod = jest.spyOn(Board.methods, 'checkWinner');
    const wrapper = mount( Board, { store, localVue });

    await wrapper.findComponent({ ref: "cell-0-0" }).trigger('click');
    expect(mockMethod).toHaveBeenCalled();
  })

  it('Should call checkStalemate when cell is clicked', async () => {
    const mockMethod = jest.spyOn(Board.methods, 'checkStalemate');
    const wrapper = mount( Board, { store, localVue });

    await wrapper.findComponent({ ref: "cell-0-0" }).trigger('click');
    expect(mockMethod).toHaveBeenCalled();
  })

  it('Should call updateScore when cell is clicked', async () => {
    const mockMethod = jest.spyOn(Board.methods, 'updateScore');
    const wrapper = mount( Board, { store, localVue });

    await wrapper.findComponent({ ref: "cell-0-0" }).trigger('click');
    expect(mockMethod).toHaveBeenCalled();
  })

  it('Should change player when cell is clicked', async () => {
    const wrapper = mount( Board, { store, localVue });

    await wrapper.findComponent({ ref: "cell-0-0" }).trigger('click');
    expect(wrapper.vm.$data.player).toBe('o');
  })  

  it('Should increment score of player1 when game is finished', async () => {
    const wrapper = mount( Board, { store, localVue });

    await wrapper.findComponent({ ref: "cell-0-0" }).trigger('click');
    await wrapper.findComponent({ ref: "cell-0-1" }).trigger('click');
    await wrapper.findComponent({ ref: "cell-1-1" }).trigger('click');
    await wrapper.findComponent({ ref: "cell-1-0" }).trigger('click');
    await wrapper.findComponent({ ref: "cell-2-2" }).trigger('click');
    
    expect(wrapper.vm.$data.finished).toBeTruthy();
    expect(wrapper.vm.$data.player1).toBe(1);
  })
})
