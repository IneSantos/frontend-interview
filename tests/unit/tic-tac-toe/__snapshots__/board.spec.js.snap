// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Board.vue renders default 1`] = `
<div class="tictactoe-board">
  <div class="grid">
    <div>
      <div>
        <cell-stub value=""></cell-stub>
      </div>
      <div>
        <cell-stub value=""></cell-stub>
      </div>
      <div>
        <cell-stub value=""></cell-stub>
      </div>
    </div>
    <div>
      <div>
        <cell-stub value=""></cell-stub>
      </div>
      <div>
        <cell-stub value=""></cell-stub>
      </div>
      <div>
        <cell-stub value=""></cell-stub>
      </div>
    </div>
    <div>
      <div>
        <cell-stub value=""></cell-stub>
      </div>
      <div>
        <cell-stub value=""></cell-stub>
      </div>
      <div>
        <cell-stub value=""></cell-stub>
      </div>
    </div>
  </div>
  <div class="score">
    <player-stub name="Player 1" score="0" class="score-info"></player-stub>
    <stopwatch-stub class="score-info"></stopwatch-stub>
    <player-stub name="Player 2" score="0" class="score-info"></player-stub>
  </div>
  <!---->
</div>
`;

exports[`Board.vue renders finished 1`] = `
<div class="tictactoe-board">
  <div class="grid">
    <div>
      <div>
        <div class="cell">
          <!---->
        </div>
      </div>
      <div>
        <div class="cell">
          <!---->
        </div>
      </div>
      <div>
        <div class="cell">
          <!---->
        </div>
      </div>
    </div>
    <div>
      <div>
        <div class="cell">
          <!---->
        </div>
      </div>
      <div>
        <div class="cell">
          <!---->
        </div>
      </div>
      <div>
        <div class="cell">
          <!---->
        </div>
      </div>
    </div>
    <div>
      <div>
        <div class="cell">
          <!---->
        </div>
      </div>
      <div>
        <div class="cell">
          <!---->
        </div>
      </div>
      <div>
        <div class="cell">
          <!---->
        </div>
      </div>
    </div>
  </div>
  <div class="score">
    <div class="player score-info">
      <h1>Player 1</h1>
      <p>0</p>
    </div>
    <div class="score-info">
      <p>00:00:00</p>
    </div>
    <div class="player score-info">
      <h1>Player 2</h1>
      <p>0</p>
    </div>
  </div>
  <div class="status">
    <div class="result"><span class="winner"></span> wins!</div>
  </div>
</div>
`;
